src/main/kotlin/org/rust/lang/core/grammar/rust.bnf

```diff
@@ -582,11 +582,13 @@ trait_ref ::= path_generic_args_without_colons
 // Impl
 ///////////////////////////////////////////////////////////////////////////////////////////////////
 
-impl_item ::= UNSAFE? IMPL generic_params type where_clause? '{' inner_attr* impl_member* '}'
-            | UNSAFE? IMPL generic_params '!'? trait_ref FOR (type where_clause? '{' inner_attr* impl_member* '}' | '..' '{' '}') {
-  pin(".*") = "'{'"
+impl_item ::= UNSAFE? IMPL generic_params type where_clause? impl_members_block
+            | UNSAFE? IMPL generic_params '!'? trait_ref FOR (type where_clause? impl_members_block | '..' '{' '}') {
+  pin(".*") = '{'
 }
 
+impl_members_block ::= '{' (inner_attr | impl_member)* '}'
```

Comment:
Let me elaborate a bit why I think handling this at the grammar level is the right way. The main reason is that it is just dramatically less code. Of course we can introduce an annotator for this. And it will be able to provide better error messages and even a quick fix to put inner attributes on top. 

But this is a lot of code which has to be maintained. And anywhere else in the project you'll have to assume that attributes and methods can be in mixed order in `impl`s. 

On the other hand the gain from the custom annotator seems pretty small: inner attributes in impls are rarely used. So spending more than a couple lines of code on this thing seems wasteful.